src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -47,6 +47,31 @@ def boolified_type(symbol_type: ct.SymbolType) -> ct.ScalarType | ct.FieldType:
     raise GTTypeError(f"Can not boolify type {symbol_type}!")
 
 
+class FieldOperatorTypeDeductionCompletnessValidator(NodeVisitor):
+    """Validate an FOAST expression is fully typed."""
+
+    @classmethod
+    def apply(cls, node: foast.LocatedNode) -> None:
+        incomplete_nodes: list[foast.LocatedNode] = []
+        cls().visit(node, incomplete_nodes=incomplete_nodes)
+
+        if incomplete_nodes:
+            raise AssertionError("FOAST expression is not fully typed.")
+
+    def visit_LocatedNode(
+        self, node: foast.LocatedNode, *, incomplete_nodes: list[foast.LocatedNode]
+    ):
+        num_incomplete_prior = len(incomplete_nodes)
+        self.generic_visit(node, incomplete_nodes=incomplete_nodes)
+
+        if (
+            len(incomplete_nodes) == num_incomplete_prior
```

Comment:
The idea is to not collect a node if one of its child node is already incomplete. With the current typing there is no case where this happens, but with parametric types this would keep the output from the pass meaningful. For example if you have `incomplete_type_expr + complete_type_expr` then the resulting type is also incomplete. However we are not interested in the parent not, but only `incomplete_type_expr`. I would like to keep it, but if you want I'll remove it.

Reply:
Ok, that was also my understanding of this condition, but the `AssertionError` in `apply()` is very confusing. I suggest to either add your explanation as a comment or delete it for now.

Reply:
I removed it.